lib/analysis/modules/ies_enkf_data.c

```diff
@@ -0,0 +1,232 @@
+   bool_vector_type * obs_mask0;      // Initial observation mask for active measurements
+   bool_vector_type * obs_mask;       // Current observation mask
+   matrix_type * W;                   // Coefficient matrix used to compute Omega = I + W (I -11'/N)/sqrt(N-1)
+   matrix_type * A0;                  // Prior ensemble used in Ei=A0 Omega_i
+   matrix_type * E;                   // Prior ensemble of measurement perturations (should be the same for all iterations)
+   bool      converged;               // GN has converged
+   ies_enkf_config_type * config;     // This I don't understand but I assume I include data from the ies_enkf_config_type defined in ies_enkf_config.c
+   bool      AAprojection;            // For including the AAprojection of Y
+   FILE*     log_fp;                  // logfile id
+};
+
+UTIL_SAFE_CAST_FUNCTION( ies_enkf_data , IES_ENKF_DATA_TYPE_ID )
+UTIL_SAFE_CAST_FUNCTION_CONST( ies_enkf_data , IES_ENKF_DATA_TYPE_ID )
+
+void * ies_enkf_data_alloc( rng_type * rng) {
+  ies_enkf_data_type * data = util_malloc( sizeof * data);
+  UTIL_TYPE_ID_INIT( data , IES_ENKF_DATA_TYPE_ID );
+  data->iteration_nr         = 0;
+  data->ies_steplength       = 0.0;
+  data->gauss_newton_conv    = 0.0;
+  data->max_gauss_newton_it  = 0;
+  data->state_size           = 0;
+  data->ens_mask0            = NULL;
+  data->ens_mask             = NULL;
+  data->obs_mask0            = NULL;
+  data->obs_mask             = NULL;
+  data->W                    = NULL;
+  data->A0                   = NULL;
+  data->E                    = NULL;
+  data->converged            = false;
+  data->config               = ies_enkf_config_alloc();
+  data->AAprojection         = true;
+  data->log_fp               = NULL;
+  return data;
+}
+
+void ies_enkf_data_free( void * arg ) {
+  ies_enkf_data_type * data = ies_enkf_data_safe_cast( arg );
+  ies_enkf_config_free( data->config );
+  free( data );
+}
+
+void ies_enkf_data_set_iteration_nr( ies_enkf_data_type * data , int iteration_nr) {
+  data->iteration_nr = iteration_nr;
+}
+
+int ies_enkf_data_inc_iteration_nr( ies_enkf_data_type * data) {
+  data->iteration_nr++;
+  return data->iteration_nr;
+}
+
+int ies_enkf_data_get_iteration_nr( const ies_enkf_data_type * data ) {
+  return data->iteration_nr;
+}
+
+
+const ies_enkf_config_type* ies_enkf_data_get_config(const ies_enkf_data_type * data) {
+  return data->config;
+}
+
+
+
+void ies_enkf_data_update_ens_mask(ies_enkf_data_type * data, const bool_vector_type * ens_mask) {
+  if (data->ens_mask)
+    bool_vector_free(data->ens_mask);
+
+  data->ens_mask = bool_vector_alloc_copy( ens_mask );
+}
+
+
+void ies_enkf_store_initial_obs_mask(ies_enkf_data_type * data, const bool_vector_type * obs_mask) {
+  if (!data->obs_mask0)
+    data->obs_mask0 = bool_vector_alloc_copy(obs_mask);
+}
+
+void ies_enkf_update_obs_mask(ies_enkf_data_type * data, const bool_vector_type * obs_mask) {
+  if (!data->obs_mask)
```

Comment:
Is this one correct
I guess the negation has to be removed.

Reply:
Yes - you are right; it should be:
```C
if (dara->obs_mask)
   bool_vector_free( data->obs_mask );
```